File: example/host_native_android_view/android/app/src/main/kotlin/dev/masel/host_native_android_view/CompassBloc.kt
package dev.masel.host_native_android_view

import com.beust.klaxon.Klaxon
import dev.masel.synced_bloc.NativeSyncSubscriber

private val klaxon = Klaxon()

class CompassBloc {
    val syncSubscriber = NativeSyncSubscriber.withMasterId("compass")
    val state: CompassState?
        get() = if (syncSubscriber.jsonState == null) null else CompassState.fromJson(syncSubscriber.jsonState!!)

    fun setDegrees(degrees: Double) {
        syncSubscriber.addEvent(CompassEventSetDegrees(degrees).toJson())
    }
}

data class CompassState (
    val degrees: Double,
    val fps: Long,
    val showMarker: Boolean
) {
    public fun toJson() = klaxon.toJsonString(this)

    companion object {
        public fun fromJson(json: String) = klaxon.parse<CompassState>(json)
    }
}

data class CompassEventSetDegrees (
    val degrees: Double,
    val runtimeType: String = "setCompassDegrees"
) {
    public fun toJson() = klaxon.toJsonString(this)

    companion object {
        public fun fromJson(json: String) = klaxon.parse<CompassEventSetDegrees>(json)
    }
}

File: example/host_native_android_view/lib/bloc/compass_bloc.dart
import 'package:flutter_bloc/flutter_bloc.dart';
import 'package:freezed_annotation/freezed_annotation.dart';
import 'package:synced_bloc/synced_bloc.dart';

part 'compass_bloc.freezed.dart';
part 'compass_bloc.g.dart';
part 'compass_event.dart';
part 'compass_state.dart';

class CompassBloc extends Bloc<CompassEvent, CompassState>
    with SyncMasterMixin {
  CompassBloc()
      : super(const CompassState(degrees: 0, fps: 60, showMarker: true)) {
    setupSyncMaster(
        masterId: 'compass',
        stateToJson: (state) => state.toJson(),
        eventFromJson: (json) => CompassEvent.fromJson(json));
    on<_SetCompassDegrees>(_onSetCompassDegrees);
    on<_SetRefreshRate>(_onSetRefreshRate);
    on<_ToggleShowMarker>(_onToggleShowMarker);
  }

  void _onSetCompassDegrees(
      _SetCompassDegrees event, Emitter<CompassState> emit) {
    emit(state.copyWith(degrees: event.degrees));
  }

  void _onSetRefreshRate(_SetRefreshRate event, Emitter<CompassState> emit) {
    int? fps = int.tryParse(event.fps);
    if (fps == null) return;
    if (fps < 1) fps = 1;
    if (fps > 200) fps = 200;
    emit(state.copyWith(fps: fps));
  }

  void _onToggleShowMarker(
      _ToggleShowMarker event, Emitter<CompassState> emit) {
    emit(state.copyWith(showMarker: !state.showMarker));
  }
}
